src/doom/v_snow.c

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright(C) 1993-1996 Id Software, Inc.
+// Copyright(C) 2005-2014 Simon Howard
+// Copyright(C) 2015-2021 Fabian Greffrath
+// Copyright(C) 2021      Mykhailo Chernysh
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	[crispy] Snow drawing
+//
+
+#include <stdlib.h>
+
+#include "v_snow.h"
+#include "i_video.h"
+#include "z_zone.h"
+#include "r_main.h"
+#include "m_random.h"
+
+typedef struct snowflake_t
+{
+    int x, y;
+} snowflake_t;
+
+static snowflake_t *snowflakes = NULL;
+static size_t snowflakes_num;
+static int snowflakes_color = -1;
+static int last_screen_size;
+static int wind = 1;
+
+static void InitSnowCoords()
+{
+    for (size_t i = 0; i < snowflakes_num; i++)
+    {
+        // here rand() but not Crispy_Random() because
+        // Crispy_Random() returns too small numbers
+        snowflakes[i].y = 0 - rand() % (SCREENHEIGHT * 2);
+        snowflakes[i].x = rand() % (SCREENWIDTH);
+    }
+}
+
+static void ResetSnow()
+{
+
+    last_screen_size = SCREENWIDTH * SCREENHEIGHT;
+    snowflakes_num = last_screen_size >> 6;
+
+    if (snowflakes != NULL)
+        free(snowflakes);
+    
+    snowflakes = malloc(snowflakes_num * sizeof(snowflake_t));
```

Comment:
No need to `free()` and `alloc()`. `realloc(NULL, ...)` is the same as `alloc(...)`.

Reply:
I think it's better to use free() cause snow animation should to be restarted when the resolution is changed.